exchange/auction_test.go

```diff
@@ -94,92 +112,143 @@ func loadCacheSpec(filename string) (*cacheSpec, error) {
-			winningBids[pbsBid.Bid.ImpID] = bid
+		cpm := bid.bid.Price
+		wbid, ok := winningBids[bid.bid.ImpID]
+		if !ok || cpm > wbid.bid.Price {
+			winningBids[bid.bid.ImpID] = bid
+		}
+		if bidMap, ok := winningBidsByImp[bid.bid.ImpID]; ok {
+			bestSoFar, ok := bidMap[pbsBid.Bidder]
+			if !ok || cpm > bestSoFar.bid.Price {
+				bidMap[pbsBid.Bidder] = bid
+			}
+		} else {
+			winningBidsByImp[bid.bid.ImpID] = make(map[openrtb_ext.BidderName]*pbsOrtbBid)
+			winningBidsByImp[bid.bid.ImpID][pbsBid.Bidder] = bid
 		}
-		winningBidsByBidder[pbsBid.Bid.ID][pbsBid.Bidder] = bid
+
+		winningBidsByImp[pbsBid.Bid.ImpID][pbsBid.Bidder] = bid
 		if len(pbsBid.Bid.Cat) == 1 {
-			bidCategory[pbsBid.Bid.ID] = pbsBid.Bid.Cat[0]
+			bidCategory[pbsBid.Bid.ImpID] = pbsBid.Bid.Cat[0]
 		}
 		roundedPrices[bid] = strconv.FormatFloat(bid.bid.Price, 'f', 2, 64)
-		// Marshal the bid for the expected cacheables
-		cjson, _ := json.Marshal(bid.bid)
-		specData.ExpectedCacheables[i].Data = cjson
 	}
 	ctx := context.Background()
 	cache := &mockCache{}
 
+	targData := &targetData{
+		priceGranularity: openrtb_ext.PriceGranularity{
+			Precision: 2,
+			Ranges: []openrtb_ext.GranularityRange{
+				{
+					Min:       0,
+					Max:       5,
+					Increment: 0.05,
+				},
+				{
+					Min:       5,
+					Max:       10,
+					Increment: 0.1,
+				},
+				{
+					Min:       10,
+					Max:       20,
+					Increment: 0.5,
+				},
+			},
+		},
+		includeWinners:    specData.TargetDataIncludeWinners,
+		includeBidderKeys: specData.TargetDataIncludeBidderKeys,
+		includeCacheBids:  specData.TargetDataIncludeCacheBids,
+		includeCacheVast:  specData.TargetDataIncludeCacheVast,
+	}
+
 	testAuction := &auction{
 		winningBids:         winningBids,
-		winningBidsByBidder: winningBidsByBidder,
+		winningBidsByBidder: winningBidsByImp,
 		roundedPrices:       roundedPrices,
 	}
-	_ = testAuction.doCache(ctx, cache, bids, vast, &specData.BidRequest, 60, &specData.DefaultTTLs, bidCategory)
-	found := 0
+	_ = testAuction.doCache(ctx, cache, targData, &specData.BidRequest, 60, &specData.DefaultTTLs, bidCategory)
 
-	for _, cExpected := range specData.ExpectedCacheables {
-		for _, cFound := range cache.items {
-			// make sure Data section matches exactly
-			if !vast {
-				// not testing VAST XML for custom cache key test
-				eq := jsonpatch.Equal(cExpected.Data, cFound.Data)
-				if !eq {
-					continue
-				}
-			}
+	if len(specData.ExpectedCacheables) > len(cache.items) {
+		//t.Errorf("%s:  Less elements were cached than expected \n", fileDisplayName)
+		t.Errorf("%s:  [CACHE_ERROR] Less elements were cached than expected \n", fileDisplayName)
+	} else if len(specData.ExpectedCacheables) < len(cache.items) {
+		t.Errorf("%s:  [CACHE_ERROR] More elements were cached than expected \n", fileDisplayName)
+	} else { // len(specData.ExpectedCacheables) == len(cache.items)
+		var ht map[string]*cacheComparator = make(map[string]*cacheComparator)
```

Comment:
Is there a reason this is a map of pointers to `cacheComparator`s rather than simply a map of `cacheComparator`s?

Reply:
As Gus found, it seems it actually needs to be a pointer for the . assignments to work properly.